Allow overriding readline completion in `auto_prepend_file`

TysonAndre opened the pull request:
Currently, it's possible to override `php -a`s completion
functionality to provide an alternative to the C implementation,
with `readline_completion_function()`.

However, that surprisingly gets overridden when called from
`auto_prepend_file`, because those scripts get run before the interactive shell
is started. I believe that not overriding it would be more consistent
with what happens when you override the completion function **after** the
interactive shell.

CLI is the only built-in API that uses this.
I believe MINIT and RINIT will only run once when invoked with `php -a`.

TysonAndre commented:
@cmb69 @nikic - I see you've contributed to readline, and it looks like readline hasn't been changed substantially since 2011 or so (e.g. 618b480d910a837c61ab554e77af45f66a8a770a changed the handling of the readline callback)

I'd consider the current behavior of readline in auto_prepend_file to be a bug (see PR description and  https://externals.io/message/111073)

What would you suggest for me to do to include changes to readline in php 8.0/8.1?

cmb69 commented:
This patch looks good to me; I suggest to merge it, if nobody objects.

However, it seems to me that `rl_attempted_completion_function` should be a module global instead of a true global.

TysonAndre commented:
> However, it seems to me that `rl_attempted_completion_function` should be a module global instead of a true global.

1. Readline is only ever used from an interactive CLI process, as far as I can tell. I can't imagine it being used in threaded mode, and anyone unexpectedly attempting to use it in embed (doubt it) probably got it working already
2. More importantly, `rl_attempted_completion_function` is a C static variable **defined by the readline C library used by php** and linked against. http://web.mit.edu/gnu/doc/html/rlman_2.html mentions that C code must override that function pointer.

   So any attempt to make this safe in threaded mode would likely fail, and setting up mutexes would be something I'd consider unexpected overkill because of the way readline is usually used (e.g. no independent threaded php processes).

johannes commented:
Correct `rl_attempted_completion_function` is exported by the deadline library, not PHP.

Please not my note above about `_readline_completion_cb`. This should get a `php` prefix or similar.

For some context: This indeed should only be used in CLI SAPI. Historically the code lived in `SAPI/cli` but since many distributions build readline as shared. Therefore I split that code partly into `ext/readline` so this can dynamically be loaded. With other SAPIs this is disabled. (I remember some weirdness around cgi, as CLI was created out of CGI, but even there it should be disabled)

cmb69 commented:
Thanks for explanaining! :)

TysonAndre commented:
> Please not my note above about `_readline_completion_cb`. This should get a php prefix or similar.

Renamed to `php_readline_completion_cb and added a README with the information mentioned here. This will get merged if nobody objects. Missing information in the README could get amended in future PRs/commits.

TysonAndre commented:
> This patch looks good to me; I suggest to merge it, if nobody objects.

Merged and added a note to `UPGRADING`, this can be amended if there are issues with the wording or the section.
